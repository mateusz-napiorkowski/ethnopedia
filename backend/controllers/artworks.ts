import { NextFunction, Request, Response } from "express"
import mongoose from "mongoose"

const asyncWrapper = require("../middleware/async")

const Artwork = require("../models/artwork")

const getAllArtworks = async (req: Request, res: Response, next: NextFunction) => {
    const page = typeof req.query.page === "string" ? parseInt(req.query.page) : 1
    const limit = typeof req.query.limit === "string" ? parseInt(req.query.limit) : 5

    const skip = (page - 1) * limit

    // const total = await Artwork.countDocuments()
    const total = 20
    // const totalPages = Math.ceil(total / limit)

    const artworks = await Artwork.find({}).skip(skip).limit(limit)
    const columnNames = Object.keys(artworks[0].toObject())

    if (columnNames.indexOf("Kategoria") != -1) {
        columnNames.splice(columnNames.indexOf("Kategoria"), 1)
        columnNames.unshift("Kategoria")
    }

    res.status(200).json({ artworks, columnNames })
}

const getArtwork = async (req: Request, res: Response, next: NextFunction) => {
    const artworkId = req.params.artworkId

    try {
        if (!mongoose.isValidObjectId(artworkId)) {
            return res.status(400).json(`Invalid artwork id: ${artworkId}`)
        }

        const artwork = await Artwork.findById(artworkId).exec()
        const columnNames = Object.keys(artwork.toObject())

        if (!artwork) {
            return res.status(404).json("Artwork not found")
        } else {
            return res.status(200).json({ artwork, columnNames })
        }

    } catch (error) {
        next(error)
    }
}

const getFilteredArtworks = async (req: Request, res: Response, next: NextFunction) => {
    try {
        const category = req.params.category
        const section = req.params.section

        const records = await Artwork.find({
            "category": category,
            "sections.sectionName": section,
        }).exec()

        return res.status(200).json(records)
        if (records.length === 0) {
            return res.status(404).json("No artworks found matching the criteria")
        } else {
            return res.status(200).json(records)
        }

    } catch (error) {
        next(error)
    }
}

const createArtwork = async (req: Request, res: Response, next: NextFunction) => {
    const title = req.body.title
    const description = req.body.description
    const img = req.body.img

    try {
        const newArtwork = await Artwork.create({
            title: title,
            description: description,
            img: img,
        })

        return res.status(201).json(newArtwork)
    } catch (error) {
        next(error)
    }
}

const deleteArtwork = async (req: Request, res: Response, next: NextFunction) => {
    const artworkId = req.params.artworkId

    try {
        if (!mongoose.isValidObjectId(artworkId)) {
            return res.status(400).json("Invalid artwork id")
        }

        const artwork = await Artwork.findById(artworkId).exec()

        if (!artwork) {
            return res.status(404).json("Artwork not found")
        }

        await artwork.remove()

        return res.sendStatus(204)
    } catch (error) {
        next(error)
    }
}


module.exports = {
    getAllArtworks,
    getArtwork,
    createArtwork,
    deleteArtwork,
    getFilteredArtworks,
}