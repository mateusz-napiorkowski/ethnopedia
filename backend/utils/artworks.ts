import { getAllCategories } from "./categories";

const fillSubcategoriesFilterPart: any = (searchText: string, currentDepth: number, maxDepth: number) => {
    if (maxDepth === 0) return []
    return {
        $elemMatch: {
            $or: currentDepth === maxDepth 
                ? [{ values: [searchText] }] 
                : [
                    { values: [searchText] },
                    { subcategories: fillSubcategoriesFilterPart(searchText, currentDepth + 1, maxDepth) }
                ]
        }
    }
}

export const constructQuickSearchFilter = async (searchText: any, collectionName: string) => {
    const allCategories = await getAllCategories(collectionName)
    const maxDepth = allCategories.length > 0 ?
        Math.max.apply(Math, allCategories.map((category) => category.split('.').length)) : 0
    const queryFilter = {
        collectionName: collectionName,
        categories: fillSubcategoriesFilterPart(searchText, 1, maxDepth)
    }
    return queryFilter
}

const constructAdvSearchSubcategoriesFilter = (searchRules: Array<Array<string>>, depth: number) => {
    let subcategoryFilter: any = {$all: []}
    searchRules.forEach(([subcategoryName, subcategoryValue]) => {
        const subcategoryNameSplitByDot = subcategoryName.split('.');
        const isCurrentDepthSubcategory = subcategoryNameSplitByDot.length === depth;
        if (!isCurrentDepthSubcategory) return

        const deeperSubcategoriesSearchRules = searchRules.filter(([deeperSubcategoryName]) => 
            deeperSubcategoryName.startsWith(`${subcategoryName}.`)
        );

        let newFilterPart: any = {
            $elemMatch: {
                name: subcategoryNameSplitByDot.slice(depth - 1).join('.'),
            }
        };

        if(subcategoryValue)
            newFilterPart.$elemMatch.values = [subcategoryValue]

        if(deeperSubcategoriesSearchRules.length != 0)
            newFilterPart.$elemMatch.subcategories = constructAdvSearchSubcategoriesFilter(deeperSubcategoriesSearchRules, depth + 1)

        subcategoryFilter.$all.push(newFilterPart)
    })
    return subcategoryFilter
}

const getOnlySearchRulesArray = (reqQuery: any) => {
    let rulesArray: any = []
    for(const categoryName in reqQuery) {
        if(!["page", "pageSize", "sortOrder", "search"].includes(categoryName)) {
            rulesArray.push([categoryName, reqQuery[categoryName]])
            const categoryNameSplitByDot = categoryName.split('.')
            while(categoryNameSplitByDot.length !== 1) {
                categoryNameSplitByDot.pop()
                if(!reqQuery.hasOwnProperty(categoryNameSplitByDot.join('.'))) {
                    rulesArray.push([categoryNameSplitByDot.join('.'), undefined])
                }
            }
        }
    }
    return rulesArray
}

export const constructAdvSearchFilter = (requestQuery: any, collectionName: string) => {
    const searchRules = getOnlySearchRulesArray(requestQuery)

    const queryFilter: any = {
        collectionName: collectionName,
        categories: {$all: []}
    }

    searchRules.forEach(([categoryName, categoryValue]: [string, any]) => {
        if(categoryName.includes('.')) return //skip subcategory search rules

        const currentCategorySubcategoriesSearchRules = searchRules.filter(([subcategoryName]: [string, any]) => 
            subcategoryName.startsWith(`${categoryName}.`)
        );

        const categoryFilter: any = { 
            $elemMatch: { 
                name: categoryName 
            }
        };

        if(categoryValue)
            categoryFilter.$elemMatch.values = [categoryValue];
        
        if(currentCategorySubcategoriesSearchRules.length > 0)
            categoryFilter.$elemMatch.subcategories = constructAdvSearchSubcategoriesFilter(currentCategorySubcategoriesSearchRules, 2);

        queryFilter.categories.$all.push(categoryFilter);
    })
    return queryFilter
}

const findMatchingCategory = (categoryName: string, record: any) => {
    if(categoryName.split(".").length === 1) {
        if(!record.subcategories)
            return record.categories.find((category: any) => category.name === categoryName);
        return record.subcategories.find((category: any) => category.name === categoryName);
    } else {
        const prefix = categoryName.split(".")[0]
        const rest = categoryName.split(".").slice(1).join(".")
        if(!record.subcategories) {
            const parent = record.categories.find((category: any) => category.name === prefix);
            if(!parent)
                return undefined
            return findMatchingCategory(rest, parent)
        }
        const parent = record.subcategories.find((category: any) => category.name === prefix);
        if(!parent)
            return undefined
        return findMatchingCategory(rest, parent)
    }
}

export const sortRecordsByCategory = (records: any, order: string) => {
    const [categoryToSortBy, ascOrDesc] = order.split('-')

    const recordAndCategoryValuePairs = records.map((record: any) => {
        const matchingCategory = findMatchingCategory(categoryToSortBy, record)
        const categoryValue = matchingCategory ? matchingCategory.values.join(", ") : null;
        return [record, categoryValue];
    })
    .sort((a: any, b: any) => {
        if (a[1] && b[1])
            return a[1].toUpperCase().localeCompare(b[1].toUpperCase())
        
        //records which have the category to sort by come first
        if (a[1]) return -1;
        if (b[1]) return 1;
    });    
    
    const sortedRecords = recordAndCategoryValuePairs.map((pair: any) => pair[0]);
    if(ascOrDesc == "desc")
        return sortedRecords.reverse()
    return sortedRecords
}