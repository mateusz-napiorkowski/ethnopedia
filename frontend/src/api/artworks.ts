import axios from "axios"
import { API_URL } from "../config"

export const getArtwork = async (id: string) => {
    return await axios.get(`${API_URL}v1/artworks/${id}`)
        .then(res => res.data)
}

export const getArtworksForPage = async (collectionIds: Array<string>, page: number, pageSize: number, sortBy: string, sortOrder: string, searchText: string | null, searchRules: any) => {
    return await axios.get(`${API_URL}v1/artworks/`, {
        params: {
            page: page,
            pageSize: pageSize,
            searchText: searchText,
            sortBy: sortBy,
            sortOrder: sortOrder,
            search: Object.entries(searchRules).length !== 0 || searchText ? true : false,
            collectionIds: collectionIds,
            ...searchRules
        }
    })
    .then(res => res.data)
}

export const createArtwork = async (collectionId: any, categories: any, file: any, jwtToken: any) => {
    const formData = new FormData();
    formData.append("files", file);
    formData.append("categories", JSON.stringify(categories));
    formData.append("collectionId", collectionId);
    const config = {
        headers: { Authorization: `Bearer ${jwtToken}` }
    };
    return await axios
        .post(`${API_URL}v1/artworks/create`, formData, config)
        .then(res => res.data)
}

export const editArtwork = async (artworkData: any, artworkId: string, jwtToken: any) => {
    const formData = new FormData();
    formData.append("file", artworkData.file);
    formData.append("categories", JSON.stringify(artworkData.categories));
    formData.append("collectionName", artworkData.collectionName);
    formData.append("fileName", artworkData.fileName)
    const config = {
        headers: {
            Authorization: `Bearer ${jwtToken}`,
        },
    };
    return await axios
        .put(`${API_URL}v1/artworks/edit/${artworkId}`, formData, config)
        .then(res => res.data);
};

export const deleteArtworks = async (artworkIds: Array<string>, jwtToken: string) => {
    const config = {
        headers: { "Authorization": `Bearer ${jwtToken}` },
        data: { ids: artworkIds}
    };
    return await axios
        .delete(`${API_URL}v1/artworks/delete`, config)
        .then(res => res.data)
}






