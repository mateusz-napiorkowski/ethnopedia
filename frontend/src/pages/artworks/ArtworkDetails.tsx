import React from "react"
import { v4 as uuidv4 } from "uuid"
import { ReactComponent as EditIcon } from "../../assets/icons/edit.svg"
import { ReactComponent as TrashBinIcon } from "../../assets/icons/trashBin.svg"

interface ArtworkDetailsProps {
    Tytuł: string;
    Artyści: string;
    Rok: string;
    collectionName: string;
    detailsToShow: { [key: string]: string };
    showStructure: boolean;
    handleEditClick: () => void;
    setShowDeleteArtworkWarning: (value: boolean) => void;
}

const ArtworkDetails: React.FC<ArtworkDetailsProps> = ({
                                                           Tytuł,
                                                           Artyści,
                                                           Rok,
                                                           collectionName,
                                                           detailsToShow,
                                                           showStructure,
                                                           handleEditClick,
                                                           setShowDeleteArtworkWarning,
                                                       }) => {
    return (
        <div className="flex flex-row">
            <div className="mt-2 flex-2">
                <h1 className="text-4xl font-bold text-gray-800 dark:text-white mb-2">{Tytuł}</h1>
                <p className="text-xl text-gray-600 dark:text-gray-400 mt-1">Artyści: {Artyści}</p>
                <p className="text-lg text-gray-500 dark:text-gray-300 mt-1">Rok: {Rok}</p>
                <p className="text-lg text-gray-500 dark:text-gray-300 mt-1">Kolekcja: {collectionName}</p>
                
                {/* {Object.entries(detailsToShow).filter(([key]) => !["__v", "userId", "_id"].includes(key)).map(([columnName, value]) => {
                    let displayValue
                    let displayName = columnName

                    if (columnName === "createdAt") {
                        displayName = "Data dodania"
                        displayValue = new Date(value).toLocaleString("pl-PL", { year: "numeric", month: "long", day: "numeric", hour: "2-digit", minute: "2-digit" })
                    } else if (columnName === "updatedAt") {
                        displayName = "Data edycji"
                        displayValue = new Date(value).toLocaleString("pl-PL", { year: "numeric", month: "long", day: "numeric", hour: "2-digit", minute: "2-digit" })
                    } else {
                        displayValue = value
                    }

                    return <div key={uuidv4()}
                                className="flex py-2 font-medium w-128">
                            <span className="mr-2 flex-1">
                                {displayName}:
                            </span>
                        <div className="flex flex-1 items-end">
                            {displayValue}
                        </div>
                    </div>
                })} */}
            </div>

            {!showStructure && (
                <div className="flex-1 mt-4 flex justify-end text-gray-700">
                    <button className="text-lg font-semibold h-fit mr-4" onClick={handleEditClick}>
                        <span className="flex-row flex items-center">
                            <EditIcon />
                            <p className="ml-2">Edytuj</p>
                        </span>
                    </button>

                    <button
                        className="text-lg font-semibold h-fit border-red-700 text-red-700 bg-red-50 hover:bg-white"
                        onClick={() => setShowDeleteArtworkWarning(true)}>
                        <span className="flex-row flex items-center">
                            <TrashBinIcon />
                            <p className="ml-2">Usuń</p>
                        </span>
                    </button>
                </div>
            )}
        </div>
    )
}
export default ArtworkDetails
