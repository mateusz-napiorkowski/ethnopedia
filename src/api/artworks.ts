import axios from "axios"

const API_URL = "http://localhost:8080/api/"
export const getArtworks = async () => {
    const response = await axios.get(`${API_URL}v1/artwork`)
    return response.data
}

export const getArtwork = async (id: any) => {
    return await axios.get(`${API_URL}v1/artwork/${id}`).then(res => res.data)
}

export const createArtwork = async () => {
    return await axios
        .post(`${API_URL}v1/artwork`, {})
        .then(res => res.data)
}


export const updatePostFn = async ({ id, artwork }: {
    id: string
    artwork: any
}) => {
    const response = await axios.patch(`v1/artwork/${id}`, artwork)
    return response.data
}

export const deletePostFn = async (id: string) => {
    const response = await axios.delete(`v1/artwork/${id}`)
    return response.data
}