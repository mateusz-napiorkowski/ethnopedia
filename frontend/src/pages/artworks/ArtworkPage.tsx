import LoadingPage from "../LoadingPage"
import Navbar from "../../components/navbar/Navbar"
import Navigation from "../../components/Navigation"
import { useState } from "react"
import WarningPopup from "../WarningPopup"
import { deleteArtworks, getArtwork } from "../../api/artworks"
import { useMutation, useQuery, useQueryClient } from "react-query"
import { useNavigate, useParams } from "react-router-dom"
import { useUser } from "../../providers/UserProvider"
import ArtworkDetails from "../../components/artwork/ArtworkDetails"

const ArtworkPage = () => {
    const { jwtToken } = useUser()
    const queryClient = useQueryClient()
    const { artworkId } = useParams<string>()
    const navigate = useNavigate()
    const [showMore, setShowMore] = useState(false)
    const [showDeleteArtworkWarning, setShowDeleteArtworkWarning] = useState(false)

    const { data: fetchedData } = useQuery({
        queryKey: ["artwork", artworkId],
        queryFn: () => getArtwork(artworkId as string),
        enabled: !!artworkId,
    })

    const findCategoryValue = (artwork: any, categoryName: string) => {
        const foundCategory = artwork.categories.find((category: any) => category.name === categoryName);
        return foundCategory ? foundCategory.value : "";
    }

    const deleteArtworksMutation = useMutation(() => deleteArtworks([artworkId as string], jwtToken as string), {
        onSuccess: () => {
            queryClient.invalidateQueries("artwork")
            navigate(`/collections/${artworkData.collectionName}/artworks/`)
        }
    })

    if (!fetchedData)
        return (
            <div data-testid="loading-page-container">
                <LoadingPage />
            </div>
        )

    const artworkData = fetchedData.artwork
    return (
        <>
            <div data-testid="loaded-artwork-page-container">
                <Navbar />
                {showDeleteArtworkWarning &&
                    <WarningPopup 
                        onClose={() => setShowDeleteArtworkWarning(!showDeleteArtworkWarning)}
                        deleteSelected={() => deleteArtworksMutation.mutate()}
                        warningMessage={"Czy na pewno chcesz usunąć rekord?"}
                    />
                }
                <section className="p-2 sm:p-4">
                    <div className="mx-auto max-w-screen-xl lg:px-6">
                        <Navigation artworkTitle={findCategoryValue(artworkData, "Tytuł")}/>
                        <ArtworkDetails
                            Tytuł={findCategoryValue(artworkData, "Tytuł")}
                            Artyści={findCategoryValue(artworkData, "Artyści")}
                            Rok={findCategoryValue(artworkData, "Rok")}
                            collectionName={artworkData.collectionName}
                            detailsToShow={showMore ? artworkData : {}}
                            handleEditClick={() => navigate(`edit-artwork`, {state:{categories: fetchedData.artwork.categories}})}
                            setShowDeleteArtworkWarning={setShowDeleteArtworkWarning}
                        />
                        <button type="button" onClick={() => setShowMore(!showMore)}
                                className="mt-4 px-4 py-2 bg-blue-500 text-white hover:bg-blue-400 font-semibold border-none">
                            {showMore ? "Pokaż mniej" : "Pokaż więcej"}
                        </button>
                    </div>
                </section>
            </div>
        </>
    )
}
export default ArtworkPage