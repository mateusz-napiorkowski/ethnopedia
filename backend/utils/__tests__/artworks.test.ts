import {describe, expect, test, jest, beforeEach} from "@jest/globals"
import { constructQuickSearchFilter, constructAdvSearchFilter, sortRecordsByCategory } from "../artworks"

const mockGetAllCategories = jest.fn()
jest.mock("../../utils/categories", () => ({
    getAllCategories: () => mockGetAllCategories(),
}))

describe('artworks util functions tests', () => {
    beforeEach(() => {
        jest.resetAllMocks()
    })

    test.each([
        {
            maxDepth: 0,
            categories: []
        },
        {
            maxDepth: 1,
            categories: ["Tytuł", "Artyści", "Rok",]
        },
        {
            maxDepth: 4,
            categories: [
                "Tytuł",
                "Tytuł.Podtytuł",
                "Tytuł.Podtytuł.Podpodtytuł",
                "Artyści",
                "Rok",
                "Rok.Miesiąc",
                "Rok.Miesiąc.Dzień",
                "Tytuł.Znaczenie tytułu",
                "Rok.Kwartał",
                "Rok.Kwartał.Miesiąc",
                "Rok.Kwartał.Miesiąc.Dzień",
              ]
        },
    ])(`constructQuickSearchFilter test - maxDepth=$maxDepth`,
        async ({categories}) => {
            mockGetAllCategories.mockReturnValue(categories)

            expect(await constructQuickSearchFilter("text to find", "collection")).toMatchSnapshot()
        }
    )

    test.each([
        {
            query: {
                page: '1',
                pageSize: '10',
                search: 'true',
                'Tytuł.Podtytuł.Podpodtytuł': 'podpodtytuł',
                'Tytuł': 'tytuł'
              }
        },
        {
            query: {
                page: '1',
                pageSize: '10',
                search: 'true',
              }
        },
    ])(`constructAdvSearchFilter test`,
        async ({query}) => {
            expect(await constructAdvSearchFilter(query, "collection")).toMatchSnapshot()
        }
    )
})