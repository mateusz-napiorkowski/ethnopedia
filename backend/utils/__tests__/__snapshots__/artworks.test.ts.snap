// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`artworks util functions tests constructAdvSearchFilter test - no search rules 1`] = `
{
  "categories": {
    "$all": [],
  },
  "collectionName": "collection",
}
`;

exports[`artworks util functions tests constructAdvSearchFilter test - standard input 1`] = `
{
  "categories": {
    "$all": [
      {
        "$elemMatch": {
          "name": "Tytuł",
          "subcategories": {
            "$all": [
              {
                "$elemMatch": {
                  "name": "Podtytuł",
                  "subcategories": {
                    "$all": [
                      {
                        "$elemMatch": {
                          "name": "Podpodtytuł",
                          "subcategories": {
                            "$all": [
                              {
                                "$elemMatch": {
                                  "name": "Podpodpodtytuł",
                                  "values": [
                                    "testowy podpodpodtytuł",
                                  ],
                                },
                              },
                            ],
                          },
                        },
                      },
                    ],
                  },
                  "values": [
                    "testowy podtytuł",
                  ],
                },
              },
            ],
          },
          "values": [
            "testowy",
          ],
        },
      },
      {
        "$elemMatch": {
          "name": "Artyści",
          "values": [
            "testowi",
          ],
        },
      },
      {
        "$elemMatch": {
          "name": "Rok",
          "subcategories": {
            "$all": [
              {
                "$elemMatch": {
                  "name": "Miesiąc",
                  "values": [
                    "Luty",
                  ],
                },
              },
            ],
          },
          "values": [
            "966",
          ],
        },
      },
    ],
  },
  "collectionName": "collection",
}
`;

exports[`artworks util functions tests constructQuickSearchFilter test - maxDepth=0 1`] = `
{
  "categories": [],
  "collectionName": "collection",
}
`;

exports[`artworks util functions tests constructQuickSearchFilter test - maxDepth=1 1`] = `
{
  "categories": {
    "$elemMatch": {
      "$or": [
        {
          "values": [
            "text to find",
          ],
        },
      ],
    },
  },
  "collectionName": "collection",
}
`;

exports[`artworks util functions tests constructQuickSearchFilter test - maxDepth=4 1`] = `
{
  "categories": {
    "$elemMatch": {
      "$or": [
        {
          "values": [
            "text to find",
          ],
        },
        {
          "subcategories": {
            "$elemMatch": {
              "$or": [
                {
                  "values": [
                    "text to find",
                  ],
                },
                {
                  "subcategories": {
                    "$elemMatch": {
                      "$or": [
                        {
                          "values": [
                            "text to find",
                          ],
                        },
                        {
                          "subcategories": {
                            "$elemMatch": {
                              "$or": [
                                {
                                  "values": [
                                    "text to find",
                                  ],
                                },
                              ],
                            },
                          },
                        },
                      ],
                    },
                  },
                },
              ],
            },
          },
        },
      ],
    },
  },
  "collectionName": "collection",
}
`;

exports[`artworks util functions tests sortRecordsByCategory test - category to sort records by doesn't exist on some records 1`] = `
[
  {
    "__v": 0,
    "_id": "6718f28bc89e4d053eebb5df",
    "categories": [
      {
        "name": "Tytuł",
        "subcategories": [],
        "values": [
          "Inny tytuł",
        ],
      },
      {
        "name": "Artyści",
        "subcategories": [],
        "values": [
          "Inny artysta",
        ],
      },
      {
        "name": "Rok",
        "subcategories": [],
        "values": [
          "1410",
        ],
      },
    ],
    "collectionName": "nowa",
    "createdAt": "2024-10-23T12:56:43.182Z",
    "updatedAt": "2024-10-23T12:56:43.182Z",
  },
  {
    "__v": 0,
    "_id": "6718f272c89e4d053eebb5d8",
    "categories": [
      {
        "name": "Tytuł",
        "subcategories": [
          {
            "name": "Podtytuł",
            "values": [
              "podtytuł",
            ],
          },
        ],
        "values": [
          "Tytułowy",
        ],
      },
      {
        "name": "Artyści",
        "subcategories": [],
        "values": [
          "Jan Nowak",
        ],
      },
      {
        "name": "Rok",
        "subcategories": [],
        "values": [
          "567",
        ],
      },
    ],
    "collectionName": "nowa",
    "createdAt": "2024-10-23T12:56:18.209Z",
    "updatedAt": "2024-10-23T12:56:18.209Z",
  },
  {
    "__v": 0,
    "_id": "6718f2bfc89e4d053eebb5e6",
    "categories": [
      {
        "name": "Artyści",
        "subcategories": [],
        "values": [
          "Pierwsi",
        ],
      },
      {
        "name": "Rok",
        "subcategories": [],
        "values": [
          "444",
        ],
      },
    ],
    "collectionName": "nowa",
    "createdAt": "2024-10-23T12:57:35.366Z",
    "updatedAt": "2024-10-23T12:57:35.366Z",
  },
  {
    "__v": 0,
    "_id": "6718f2bfc89e4d053eebb5e6",
    "categories": [
      {
        "name": "Artyści",
        "subcategories": [],
        "values": [
          "Pierwsi",
        ],
      },
      {
        "name": "Rok",
        "subcategories": [],
        "values": [
          "444",
        ],
      },
    ],
    "collectionName": "nowa",
    "createdAt": "2024-10-23T12:57:35.366Z",
    "updatedAt": "2024-10-23T12:57:35.366Z",
  },
]
`;

exports[`artworks util functions tests sortRecordsByCategory test - nonexistent category to sort by 1`] = `
[
  {
    "__v": 0,
    "_id": "6718f272c89e4d053eebb5d8",
    "categories": [
      {
        "name": "Tytuł",
        "subcategories": [
          {
            "name": "Podtytuł",
            "values": [
              "podtytuł",
            ],
          },
        ],
        "values": [
          "Tytułowy",
        ],
      },
      {
        "name": "Artyści",
        "subcategories": [],
        "values": [
          "Jan Nowak",
        ],
      },
      {
        "name": "Rok",
        "subcategories": [],
        "values": [
          "567",
        ],
      },
    ],
    "collectionName": "nowa",
    "createdAt": "2024-10-23T12:56:18.209Z",
    "updatedAt": "2024-10-23T12:56:18.209Z",
  },
  {
    "__v": 0,
    "_id": "6718f28bc89e4d053eebb5df",
    "categories": [
      {
        "name": "Tytuł",
        "subcategories": [],
        "values": [
          "Inny tytuł",
        ],
      },
      {
        "name": "Artyści",
        "subcategories": [],
        "values": [
          "Inny artysta",
        ],
      },
      {
        "name": "Rok",
        "subcategories": [],
        "values": [
          "1410",
        ],
      },
    ],
    "collectionName": "nowa",
    "createdAt": "2024-10-23T12:56:43.182Z",
    "updatedAt": "2024-10-23T12:56:43.182Z",
  },
  {
    "__v": 0,
    "_id": "6718f2bfc89e4d053eebb5e6",
    "categories": [
      {
        "name": "Tytuł",
        "subcategories": [
          {
            "name": "Podtytuł",
            "values": [
              "Jakiś podtytuł",
            ],
          },
        ],
        "values": [
          "Kolejny tytuł",
        ],
      },
      {
        "name": "Artyści",
        "subcategories": [],
        "values": [
          "Pierwsi",
        ],
      },
      {
        "name": "Rok",
        "subcategories": [],
        "values": [
          "444",
        ],
      },
    ],
    "collectionName": "nowa",
    "createdAt": "2024-10-23T12:57:35.366Z",
    "updatedAt": "2024-10-23T12:57:35.366Z",
  },
]
`;

exports[`artworks util functions tests sortRecordsByCategory test - sort by Artyści asc 1`] = `
[
  {
    "__v": 0,
    "_id": "6718f28bc89e4d053eebb5df",
    "categories": [
      {
        "name": "Tytuł",
        "subcategories": [],
        "values": [
          "Inny tytuł",
        ],
      },
      {
        "name": "Artyści",
        "subcategories": [],
        "values": [
          "Inny artysta",
        ],
      },
      {
        "name": "Rok",
        "subcategories": [],
        "values": [
          "1410",
        ],
      },
    ],
    "collectionName": "nowa",
    "createdAt": "2024-10-23T12:56:43.182Z",
    "updatedAt": "2024-10-23T12:56:43.182Z",
  },
  {
    "__v": 0,
    "_id": "6718f272c89e4d053eebb5d8",
    "categories": [
      {
        "name": "Tytuł",
        "subcategories": [
          {
            "name": "Podtytuł",
            "values": [
              "podtytuł",
            ],
          },
        ],
        "values": [
          "Tytułowy",
        ],
      },
      {
        "name": "Artyści",
        "subcategories": [],
        "values": [
          "Jan Nowak",
        ],
      },
      {
        "name": "Rok",
        "subcategories": [],
        "values": [
          "567",
        ],
      },
    ],
    "collectionName": "nowa",
    "createdAt": "2024-10-23T12:56:18.209Z",
    "updatedAt": "2024-10-23T12:56:18.209Z",
  },
  {
    "__v": 0,
    "_id": "6718f2bfc89e4d053eebb5e6",
    "categories": [
      {
        "name": "Tytuł",
        "subcategories": [
          {
            "name": "Podtytuł",
            "values": [
              "Jakiś podtytuł",
            ],
          },
        ],
        "values": [
          "Kolejny tytuł",
        ],
      },
      {
        "name": "Artyści",
        "subcategories": [],
        "values": [
          "Pierwsi",
        ],
      },
      {
        "name": "Rok",
        "subcategories": [],
        "values": [
          "444",
        ],
      },
    ],
    "collectionName": "nowa",
    "createdAt": "2024-10-23T12:57:35.366Z",
    "updatedAt": "2024-10-23T12:57:35.366Z",
  },
]
`;

exports[`artworks util functions tests sortRecordsByCategory test - sort by Tytuł asc 1`] = `
[
  {
    "__v": 0,
    "_id": "6718f28bc89e4d053eebb5df",
    "categories": [
      {
        "name": "Tytuł",
        "subcategories": [],
        "values": [
          "Inny tytuł",
        ],
      },
      {
        "name": "Artyści",
        "subcategories": [],
        "values": [
          "Inny artysta",
        ],
      },
      {
        "name": "Rok",
        "subcategories": [],
        "values": [
          "1410",
        ],
      },
    ],
    "collectionName": "nowa",
    "createdAt": "2024-10-23T12:56:43.182Z",
    "updatedAt": "2024-10-23T12:56:43.182Z",
  },
  {
    "__v": 0,
    "_id": "6718f2bfc89e4d053eebb5e6",
    "categories": [
      {
        "name": "Tytuł",
        "subcategories": [
          {
            "name": "Podtytuł",
            "values": [
              "Jakiś podtytuł",
            ],
          },
        ],
        "values": [
          "Kolejny tytuł",
        ],
      },
      {
        "name": "Artyści",
        "subcategories": [],
        "values": [
          "Pierwsi",
        ],
      },
      {
        "name": "Rok",
        "subcategories": [],
        "values": [
          "444",
        ],
      },
    ],
    "collectionName": "nowa",
    "createdAt": "2024-10-23T12:57:35.366Z",
    "updatedAt": "2024-10-23T12:57:35.366Z",
  },
  {
    "__v": 0,
    "_id": "6718f272c89e4d053eebb5d8",
    "categories": [
      {
        "name": "Tytuł",
        "subcategories": [
          {
            "name": "Podtytuł",
            "values": [
              "podtytuł",
            ],
          },
        ],
        "values": [
          "Tytułowy",
        ],
      },
      {
        "name": "Artyści",
        "subcategories": [],
        "values": [
          "Jan Nowak",
        ],
      },
      {
        "name": "Rok",
        "subcategories": [],
        "values": [
          "567",
        ],
      },
    ],
    "collectionName": "nowa",
    "createdAt": "2024-10-23T12:56:18.209Z",
    "updatedAt": "2024-10-23T12:56:18.209Z",
  },
]
`;

exports[`artworks util functions tests sortRecordsByCategory test - sort by Tytuł desc 1`] = `
[
  {
    "__v": 0,
    "_id": "6718f272c89e4d053eebb5d8",
    "categories": [
      {
        "name": "Tytuł",
        "subcategories": [
          {
            "name": "Podtytuł",
            "values": [
              "podtytuł",
            ],
          },
        ],
        "values": [
          "Tytułowy",
        ],
      },
      {
        "name": "Artyści",
        "subcategories": [],
        "values": [
          "Jan Nowak",
        ],
      },
      {
        "name": "Rok",
        "subcategories": [],
        "values": [
          "567",
        ],
      },
    ],
    "collectionName": "nowa",
    "createdAt": "2024-10-23T12:56:18.209Z",
    "updatedAt": "2024-10-23T12:56:18.209Z",
  },
  {
    "__v": 0,
    "_id": "6718f2bfc89e4d053eebb5e6",
    "categories": [
      {
        "name": "Tytuł",
        "subcategories": [
          {
            "name": "Podtytuł",
            "values": [
              "Jakiś podtytuł",
            ],
          },
        ],
        "values": [
          "Kolejny tytuł",
        ],
      },
      {
        "name": "Artyści",
        "subcategories": [],
        "values": [
          "Pierwsi",
        ],
      },
      {
        "name": "Rok",
        "subcategories": [],
        "values": [
          "444",
        ],
      },
    ],
    "collectionName": "nowa",
    "createdAt": "2024-10-23T12:57:35.366Z",
    "updatedAt": "2024-10-23T12:57:35.366Z",
  },
  {
    "__v": 0,
    "_id": "6718f28bc89e4d053eebb5df",
    "categories": [
      {
        "name": "Tytuł",
        "subcategories": [],
        "values": [
          "Inny tytuł",
        ],
      },
      {
        "name": "Artyści",
        "subcategories": [],
        "values": [
          "Inny artysta",
        ],
      },
      {
        "name": "Rok",
        "subcategories": [],
        "values": [
          "1410",
        ],
      },
    ],
    "collectionName": "nowa",
    "createdAt": "2024-10-23T12:56:43.182Z",
    "updatedAt": "2024-10-23T12:56:43.182Z",
  },
]
`;
